### PR TITLE
configure dedicated db user for database migrations: DML-only user fo…

### PR DESCRIPTION
**Pull request by mfpc**
…r datasource, but DDL user for migration #24639

### CODE REVIEW

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/DataSourceDecorator.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    public Connection getConnection() throws SQLException {
+        // TODO Auto-generated method stub
+        return dataSource.getConnection(username.get(), password.get());
+    }
+
+    public Connection getConnection(String username, String password)
+            throws SQLException {
+        // TODO Auto-generated method stub
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/DataSourceDecorator.java:68]: Please remove these comments

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java [hunk 2]
```diff
@@ -33,7 +33,13 @@ public FlywayCreator withCallbacks(Collection<Callback> callbacks) {
 
     public Flyway createFlyway(DataSource dataSource) {
         FluentConfiguration configure = Flyway.configure();
-        configure.dataSource(dataSource);
+
+        if (flywayRuntimeConfig.username.isPresent() && flywayRuntimeConfig.password.isPresent()) {
+            configure.dataSource(
+                    new DataSourceDecorator(dataSource, flywayRuntimeConfig.username, flywayRuntimeConfig.password));
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java:39]: You can invoke `get()` here and avoid using `Optional` in the class constructor

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/DataSourceDecorator.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import java.io.PrintWriter;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.Optional;
+import java.util.logging.Logger;
+
+import javax.sql.DataSource;
+
+public class DataSourceDecorator implements DataSource {
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/DataSourceDecorator.java:12]: The name doesn't really explain what this class does, but I am not very good at naming, so I'll let others decide about it

File: extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionDifferentUsernameTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+        int var = 0;
+        try (Connection con = datasource.getConnection();
+                PreparedStatement ps = con.prepareStatement("SELECT 100");
+                ResultSet rs = ps.executeQuery()) {
+            if (rs.next()) {
+                var = rs.getInt(1);
+            }
+        }
+        assertEquals(100, var, "Different username and password was not executed");
```
Comment (gastaldi) [inline extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionDifferentUsernameTest.java:42]: This test doesn't really verify the feature you provided

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/DataSourceDecorator.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        // TODO Auto-generated method stub
+        dataSource.setLoginTimeout(arg0);
+    }
+
+    public boolean isWrapperFor(Class<?> arg0) throws SQLException {
+        // TODO Auto-generated method stub
+        return dataSource.isWrapperFor(arg0);
+    }
+
+    public <T> T unwrap(Class<T> arg0) throws SQLException {
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/DataSourceDecorator.java:56]: Replace `arg0` parameters with the name used in the sources

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java [hunk 6]
```diff
@@ -43,6 +43,26 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
     @ConfigItem
     public Optional<String> defaultSchema = Optional.empty();
 
+    /**
+     * This is a safety measurement against theoretical situations where the application goes haywire for any reason.
+     * This includes e.g. accidentally setting quarkus.hibernate-orm.database.generation to drop-and-create.
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:48]: This description looks misleading and confusing

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java [hunk 7]
```diff
@@ -43,6 +43,26 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
...
+     * quarkus.flyway.username=my_schema_owner
+     *
+     */
+
+    @ConfigItem
+    public Optional<String> username = Optional.empty();
+
+    /**
+     * Set the password for connection;
+     * quarkus.flyway.password=my_secret_password
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:60]: This line can be removed

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java [hunk 8]
```diff
@@ -43,6 +43,26 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
     @ConfigItem
     public Optional<String> defaultSchema = Optional.empty();
 
+    /**
+     * This is a safety measurement against theoretical situations where the application goes haywire for any reason.
+     * This includes e.g. accidentally setting quarkus.hibernate-orm.database.generation to drop-and-create.
+     * If the application runs with a restricted database user, damage in such situations can be mitigated.
+     * # Name of a custom database user for flyway to use. Must have DDL rights
+     * quarkus.flyway.username=my_schema_owner
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:51]: This can be removed

**Review by gastaldi — CHANGES_REQUESTED** [review]
Added some comments. The implementation looks good.

File: extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionDifferentUsernameTest.java (hunk 4 above)
Comment (mfpc) [reply extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionDifferentUsernameTest.java:42]: This test if the connection is working with new configs.

Comment (mfpc) [reply extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionDifferentUsernameTest.java:42]: I will change the message for "Database connected with new config", its ok?

**Comment by gastaldi** [comment]
For the record, there are some worthy comments in the previous PR: https://github.com/quarkusio/quarkus/pull/29628

File: extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionDifferentUsernameTest.java (hunk 4 above)
Comment (gsmet) [reply extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionDifferentUsernameTest.java:42]: I think you should rather get the datasource and check the user there. In your previous PR, I outlined how you could get the datasource URL. I'm pretty sure you can get the user from there too.

**Review by gsmet — CHANGES_REQUESTED** [review]
Not completely sold by this new approach and I would like more details about why you changed your approach.
Have you seen my comments in your previous PR?

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java (hunk 2 above)
Comment (gastaldi) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java:39]: @gsmet Before I read your comments on the previous PR, I suggested @mfpc to use this pattern. The idea was to reuse the DataSource configuration and call `DataSource#getConnection(String user, String password)` when `DataSource#getConnection()` is called in the Flyway codebase. 

But now I see that your suggestion seems nicer, so 👍🏻 to that

**Comment by geoand** [comment]
https://github.com/quarkusio/quarkus/pull/29704 seems to be a follow up to this.

**Pull request closed without merging**